lib/internal/crypto/random.js

```diff
@@ -281,10 +282,86 @@ function getRandomValues(data) {
   return data;
 }
 
+// Implements an RFC 4122 version 4 random UUID.
+// To improve performance, random data is generated in batches
+// large enough to cover kBatchSize UUID's at a time. The uuidData
+// and uuid buffers are reused for each. Each call to randomUUID()
+// consumes 16 bytes from the buffer.
+
+const kHexDigits = [
+  48, 49, 50, 51, 52, 53, 54, 55,
+  56, 57, 97, 98, 99, 100, 101, 102
+];
+
+const kBatchSize = 128;
+let uuidData;
+let uuid;
+let uuidBatch = 0;
+
+function randomUUID() {
+  if (uuidData === undefined) {
+    uuidData = new Uint8Array(16 * kBatchSize);
+    uuid = Buffer.alloc(36, '-');
+    uuid[14] = 52; // '4'
+  }
+
+  if (uuidBatch === 0) randomFillSync(uuidData);
```

Comment:
This will lead to false positive detection when `--trace-sync-io` is in use. It's worth it to at least mention this fact in the doc.

Reply:
All other random functions in the crypto module provide asynchronous implementations as well, maybe that should be considered here, too?

Reply:
@tniessen ... Given that there's not yet an async version in the standard proposal I'd rather hold off on that. Also, there does not appear to be a perf benefit to an async variant.

Reply:
@jasnell what about the original `--trace-sync-io` concern mentioned here?

Reply:
@puzpuzpuz ... I'm not clear on what concern you're referring to.

Reply:
I'm referring to the following flag: https://nodejs.org/api/cli.html#cli_trace_sync_io

I know that users enable this flag in order to detect `Sync` operation usage after the first event loop tick. With the proposed implementation there will be false positive warnings printed into stdout when `randomUUID` is called (because of the underlying `randomFillSync`). As the operation is not marked as `Sync` and doesn't have an async version, users might be confused and start thinking that the event loop gets blocked and they shouldn't use this function.

We should at least mention this fact in the documentation to avoid user confusion. Going a step further, it might be a good idea to completely disable the check for `randomUUID` by introducing an internal parameter of `RandomBytesJob`. When this parameter is initialized, the following check should not happen: https://github.com/nodejs/node/blob/master/src/crypto/crypto_util.h#L339

BTW `uuidjs` suffers from exactly the same problem, but it's much easier to fix the problem now, as we're considering a core API.

Reply:
I'm not really convinced that it's an issue. `--trace-sync-io` is there there to let folks know when they are using a sync api, and this would qualify as that. I don't consider it a false positive or a problem

Reply:
Yes, it's technically a blocking API, but in reality each `randomUUID` should never take more than several microseconds. That's why there is no need for an async version of the function, at least at the moment. So, I still think that seeing the warnings will confuse users (and I saw that happening before with `uuidjs`).

Nevertheless, let's mark this one as resolved and wait for some feedback from users.